### PR TITLE
[node] Add crypto.KeyObject support (added in node 11.6.0)

### PR DESCRIPTION
**Pull request by btoueg**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/docs/latest-v11.x/api/crypto.html#crypto_class_keyobject
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by SimonSchick — CHANGES_REQUESTED** [review]
The features you added were specifically added in node v11, you'll have to create a new definitions folder, HOWEVER there is currently some large scale refactoring going on so I would ask to coordinate with @rbuckton here.

**Comment by btoueg** [comment]
Hi @rbuckton 👋 What's that refactoring mentioned by @SimonSchick ?

The change in TS definition in this PR is backward compatible with node 10, even though to add definitions that are not present in node 10...

What should be done here?

File: types/node/index.d.ts [hunk 1]
```diff
@@ -6776,19 +6801,39 @@ declare module "crypto" {
...
+        sign(private_key: PrivateKeyInputForSigning | string | Buffer | KeyObject): Buffer;
+        sign(private_key: PrivateKeyInputForSigning | string | Buffer | KeyObject, output_format: HexBase64Latin1Encoding): string;
     }
     function createVerify(algorith: string, options?: stream.WritableOptions): Verify;
     interface Verify extends NodeJS.WritableStream {
         update(data: string | Buffer | NodeJS.TypedArray | DataView): Verify;
         update(data: string, input_encoding: Utf8AsciiLatin1Encoding): Verify;
-        verify(object: string | Object, signature: Buffer | NodeJS.TypedArray | DataView): boolean;
-        verify(object: string | Object, signature: string, signature_format: HexBase64Latin1Encoding): boolean;
+        verify(object: Object | string | Buffer | KeyObject, signature: Buffer | NodeJS.TypedArray | DataView): boolean;
```
[the remark this answers is not in the document]
Comment (btoueg) [reply types/node/index.d.ts]: There is room for improvement, but let's agree that it's better typed than before, and also it's ordered according to the doc

**Comment by btoueg** [comment]
👋 @SimonSchick 
I've updated the PR to match the refacto, and included most of your feedback
Let's consider merging this PR as it's definitely an improvement even if it's a small incremental change

**Comment by SimonSchick** [comment]
You'll still have to move all current typings into a `v10` folder and tag these typings for node v11 in the `index.d.ts`

**Comment by Flarna** [comment]
> Even if they are backward compatible? ok

It's not only about backward compatibility. If this is added users of NodeJs 10 - which are a lot as this is the current LTS version - could call an API which doesn't exist.
By adding a dedicated v10/v11 version users can choose the matching version.

**Comment by btoueg** [comment]
I've bumped the version in the header, not sure if I should have started from 11.0 or be aligned with the current version of nodejs
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32597/files#diff-7d84e08967cded0b99ed4328aab0a1a8R1

**Comment by SimonSchick** [comment]
You should only bump it up to the version your changes actually support.

**Comment by rbuckton** [comment]
If @btoueg's changes go in before #32691, that's fine. I can update my PR and add a "typesVersions" entry for v10 as well.

File: types/node/crypto.d.ts [hunk 2]
```diff
@@ -46,8 +46,21 @@ declare module "crypto" {
...
         digest(encoding: HexBase64Latin1Encoding): string;
     }
+
+    interface KeyObject {
+        asymmetricKeyType?: KeyType;
+        export(options?: {
+            type: 'pkcs1' | 'spki' | 'pkcs8' | 'sec1',
+            format: KeyFormat,
+            cipher?: string,
+            passphrase?: string
```
Comment (Flarna) [inline types/node/crypto.d.ts]: I think passphrase can be `string | Buffer` (see https://nodejs.org/docs/latest-v11.x/api/crypto.html#crypto_keyobject_export_options)

File: types/node/crypto.d.ts [hunk 3]
```diff
@@ -46,8 +46,21 @@ declare module "crypto" {
...
+        }): string | Buffer;
+        symmetricSize: number;
```
Comment (Flarna) [inline types/node/crypto.d.ts]:
```suggestion
        symmetricSize?: number;
```
see https://nodejs.org/docs/latest-v11.x/api/crypto.html#crypto_keyobject_symmetrickeysize

File: types/node/crypto.d.ts [hunk 4]
```diff
@@ -128,19 +153,39 @@ declare module "crypto" {
         setAAD(buffer: Buffer | NodeJS.TypedArray | DataView, options?: { plaintextLength: number }): this;
     }
 
+    interface PrivateKeyInput {
+        key: string | Buffer;
+        format?: KeyFormat;
+        type?: 'pkcs1' | 'spki';
```
Comment (Flarna) [inline types/node/crypto.d.ts]:
```suggestion
        type?: 'pkcs1' | 'pkcs8' | 'sec1';
```
see https://nodejs.org/docs/latest-v11.x/api/crypto.html#crypto_crypto_createprivatekey_key

File: types/node/crypto.d.ts [hunk 5]
```diff
@@ -128,19 +153,39 @@ declare module "crypto" {
...
+    }
+
+    function createPrivateKey(key: PrivateKeyInput | string | Buffer): KeyObject;
+    function createPublicKey(key: PublicKeyInput | string | Buffer): KeyObject;
+    function createSecretKey(key: Buffer): KeyObject;
     function createSign(algorithm: string, options?: stream.WritableOptions): Signer;
+
+    interface SignPrivateKeyInput extends PrivateKeyInput {
+        saltLength?: number;
```
Comment (Flarna) [inline types/node/crypto.d.ts]: I think `padding?: number` is missing here.

**Comment by btoueg** [comment]
Thanks for the feedback @Flarna

**Comment by Flarna** [comment]
Seems there are two PRs now to create a v10 folder and add v11 changes (#32878).
@SimonSchick and @btoueg I think you should agree on who waits or you combine the changes into one PR.

**Comment by SimonSchick** [comment]
If @btoueg is ok with that I will add their changes to my PR.

**Comment by btoueg** [comment]
Go ahead

**Pull request closed without merging**